File: solutions/1036-rotting-oranges/solution.py
class Solution(object):
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """

        dirs = [[1,0],[0,1],[-1,0],[0,-1]]
        rows = len(grid)
        cols = len(grid[0])

        time = [0]
        fresh = [0]
        q = deque()

        # Initial scan
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    q.append((r, c))
                elif grid[r][c] == 1:
                    fresh[0] += 1

        seen = [[False for _ in range(cols)] for _ in range(rows)]

        
        while q and fresh[0] > 0:
            size = len(q)
            for _ in range(size):
                ro, co = q.popleft()
                for dr, dc in dirs:
                    nr, nc = ro + dr, co + dc

                    if nr < 0 or nc < 0 or nr >= rows or nc >= cols:
                        continue
                    if grid[nr][nc] != 1 or seen[nr][nc]:
                        continue

                    seen[nr][nc] = True
                    grid[nr][nc] = 2
                    fresh[0] -= 1
                    q.append((nr, nc))

            time[0] += 1

        return time[0] if fresh[0] == 0 else -1

File: solutions/1036-rotting-oranges/solution.java
import java.util.*;

class Solution {
    public int orangesRotting(int[][] grid) {
        
        int m = grid.length;
        int n = grid[0].length;
        
        Deque<int[]> q = new ArrayDeque<>();
        
        int fresh = 0;
        int time = 0;
        
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                
                if(grid[i][j]==2){
                    q.add(new int[]{i,j});
                }
                
                if(grid[i][j]==1){
                    fresh++;
                }
            }
        }
        
        if(fresh==0) return 0;
        
        int dirs[][]={{1,0},{-1,0},{0,1},{0,-1}};
        
        while(!q.isEmpty()){
            
            int size = q.size();
            boolean rotted = false;
            
            for(int k=0;k<size;k++){
                
                int cur[] = q.poll();
                
                for(int d[]:dirs){
                    
                    int nr = cur[0]+d[0];
                    int nc = cur[1]+d[1];
                    
                    if(nr<0 || nc<0 || nr>=m || nc>=n){
                        continue;
                    }
                    
                    if(grid[nr][nc]==1){
                        
                        grid[nr][nc]=2;
                        fresh--;
                        rotted = true;
                        
                        q.add(new int[]{nr,nc});
                    }
                }
            }
            
            if(rotted) time++;
        }
        
        if(fresh>0) return -1;
        
        return time;
    }
}
